packages/go_router/lib/src/route.dart

```diff
@@ -449,19 +460,296 @@ class ShellRoute extends RouteBase {
+  /// [ShellRouteBranch]s. Each GoRoute provides the navigator key
+  /// (via [GoRoute.parentNavigatorKey]) that will be used to create the
+  /// separate [Navigator]s for the routes.
+  StatefulShellRoute.rootRoutes({
+    required List<GoRoute> routes,
+    required ShellRouteBuilder builder,
+    ShellRoutePageBuilder? pageProvider,
+  }) : this(
+          branches: routes.map((GoRoute e) {
+            final GlobalKey<NavigatorState>? key = e.parentNavigatorKey;
+            assert(key != null, 'Each route must specify a parentNavigatorKey');
+            return ShellRouteBranch(rootRoute: e, navigatorKey: key!);
+          }).toList(),
+          builder: builder,
+          pageProvider: pageProvider,
+        );
+
+  /// Representations of the different stateful route branches that this
+  /// shell route will manage. Each branch identifies the [Navigator] to be used
+  /// (via the navigatorKey) and the route that will be used as the root of the
+  /// route branch.
+  final List<ShellRouteBranch> branches;
+
+  /// The widget builder for a stateful shell route.
+  ///
+  /// Similar to GoRoute builder, but with an additional child parameter. This
+  /// child parameter is a Widget that contains - and is responsible for
+  /// managing - the Navigators for the different route branches of this
+  /// StatefulShellRoute. This widget is meant to be used as the body of a
+  /// custom shell implementation, for example as the body of [Scaffold] with a
+  /// [BottomNavigationBar].
+  ///
+  /// The builder function of a StatefulShellRoute will be invoked from within a
+  /// wrapper Widget that provides access to the current
+  /// [StatefulShellRouteState] associated with the route (via the method
+  /// [StatefulShellRoute.of]). That state object exposes information such as
+  /// the current branch index, the state of the route branches etc.
+  ///
+  /// For implementations where greater control is needed over the layout and
+  /// animations of the Navigators, the child parameter can be ignored, and a
+  /// custom implementation can instead be built by using the Navigators (and
+  /// other information from StatefulShellRouteState) directly. For example:
+  ///
+  /// ```
+  /// final StatefulShellRouteState shellState = StatefulShellRoute.of(context);
+  /// final int currentIndex = shellState.currentBranchIndex;
+  /// final List<Widget?> navigators = shellRouteState.branchNavigators;
+  /// return MyCustomShell(currentIndex, navigators);
+  /// ```
+  final ShellRouteBuilder builder;
+
+  /// Function for customizing the [Page] for this stateful shell.
+  ///
+  /// Similar to GoRoute pageBuilder, but with an additional child parameter.
+  /// Unlike GoRoute however, this function is used in combination with
+  /// [builder], and the child parameter will be the stateful shell already
+  /// built for this route, using the builder function.
+  final ShellRoutePageBuilder? pageProvider;
```

Comment:
why not call this pageBuilder, it seems to be more aligned with other route type

Reply:
I wanted to differentiate it a bit from the pageBuilder used on for instance `ShellRoute`, since it works a bit different, in that it isn't a replacement for the `builder`, but used in addition to it. But I guess it's still a page builder, and after all, the documentation (and constructor requirements) describes how it should be used. So I'll change it to pageBuilder.